File: dotnet/Allmark/Render/ConsoleDeletionRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleDeletionRenderer
{
    public static Renderer Create(string style, string reset)
    {
        return new Renderer
        {
            Name = "deletion",
            Render = (node, state, first, last, decode) => Render(node, state, style, reset),
        };
    }

    public static Renderer Create()
    {
        return new Renderer
        {
            Name = "deletion",
            Render = (node, state, first, last, decode) => Render(node, state, RenderToConsole.AnsiRed, RenderToConsole.AnsiReset),
        };
    }

    public static void Render(MarkdownNode node, RendererState state, string style, string reset)
    {
        state.Output.Append($"{style}--");
        RenderChildren.Execute(node, state);
        state.Output.Append($"--{reset}");
    }
}

File: zig/src/render-console/consoleDeletionRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const ConsoleRendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const ansiDim = @import("./renderToConsole.zig").ansiDim;
const ansiStrikethrough = @import("./renderToConsole.zig").ansiStrikethrough;
const ansiStrikethroughReset = @import("./renderToConsole.zig").ansiStrikethroughReset;
const ansiReset = @import("./renderToConsole.zig").ansiReset;
const renderChildrenConsole = @import("./renderToConsole.zig").renderChildrenConsole;

pub const consoleDeletionRenderer = Renderer{
    .name = "deletion",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *ConsoleRendererState, first: ?bool, last: ?bool, decode: ?bool) void {
    _ = first;
    _ = last;
    _ = decode;

    state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
    state.output.appendSlice(state.allocator, ansiStrikethrough) catch unreachable;
    renderChildrenConsole(node, state, true) catch unreachable;
    state.output.appendSlice(state.allocator, ansiStrikethroughReset) catch unreachable;
    state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
}

pub const deletionRenderer = struct {
    name: []const u8 = "deletion",
    render: consoleDeletionRenderer,
};
